proposals/text/5540-llm-support-2.0.md

```diff
@@ -0,0 +1,309 @@
+		component.set_input_parameters(**{var: Any for var in template_variables})
+
+  	@component.output_types(prompts=List[str])
+    def run(self, template: str, **kwargs):
+        self.template_text = # Load the template
+
+        variables = # extracts the variables from the template text
+        if variables != self.template_variables:
+            raise ValueError()
+
+        # Render the template using the variables
+        return {"prompts": prompts}
+```
+
+### **Why we need to specify the variables in `__init__`?**
+
+The design above derives from one Canals limitation: component’s sockets need to be all known the latest at `__init__` time, in order for the connections to be made and validated. Therefore, we need to know all the prompt variables before building the pipelines, because the prompt variables are inputs of the `run()` method.
+
+However, earlier iterations of Canals did support so-called “true variadic” components: components that do not need to know what they will be connected to, and build the input sockets at need. Such components of course lack input validation, but enable usecases like the above.
+
+If we decide that Canals should support again such components, we would be able to rewrite `Prompts` to take a prompt as its input parameter and just accept any other incoming input, on the assumption that users knows that they’re doing.
+
+For example:
+
+```python
+@component
+class Prompts:
+
+	@variadic_input
+  	@component.output_types(prompts=List[str])
+    def run(self, template: Union[str, Path], **kwargs):
+	    # ... loads the template ...
+        # ... render the prompts ...
+        return {"prompts": prompts}
+```
+
+### Why a separate `Prompts` component at all?
+
+`PromptNode` used to take the prompt template and the variables to render it directly, and then forward the result to the LLM.
+
+**PRO: Allows using various templating and guidance libraries, or none at all**
+
+The key advantage of `Prompts` is ability to use any tool from the ever growing list of LLM prompting template libs. These include but are not limited to: https://github.com/microsoft/guidance,
+https://shreyar.github.io/guardrails/rail/prompt/
+and many many more. See list at https://www.promptingguide.ai/tools If someone has invested a lot in guidance and considers using a framework like Haystack or LangChain this will be one of the biggest selling points: an ability to use `Prompts` for a specific prompt tooling lib.
+
+On top of that, it also allows users to skip the template rendering step altogether and send prompts directly to the LLM, which may be beneficial in some context (for example, if users just want to chat with the LLM without RAG).
+
+**PRO: Abstracts away duplicate work (which may be expensive)**
+
+On top of this, there is also the question of duplication of work. In a situation where a Pipeline sends the same prompt to several LLMs, the v2 version of such pipeline can easily become very complicated. Take this usecase:
```

Comment:
I would question how often it happens that the same prompt is used multiple times. On the other hand, if prompts are tokenizer aware, then the re-use of prompts for multiple LLMs would be limited to LLMs of the same type anyways. Not sure how useful it is then.

Reply:
(as above)

Reply:
I'll resolve this one for now as I will reconsider this section entirely.